bundles/config/org.eclipse.smarthome.config.discovery.usbserial.linux.sysfs/src/main/java/org/eclipse/smarthome/config/discovery/usbserial/linux/sysfs/internal/SysfsUsbSerialScanner.java

```diff
@@ -0,0 +1,269 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.config.discovery.usbserial.linux.sysfs.internal;
+
+import static java.nio.file.Files.*;
+
+import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.discovery.usbserial.UsbSerialDeviceInformation;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Modified;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Sets;
+
+/**
+ * A {@link UsbSerialScanner} that scans the system for USB devices which provide a serial port by inspecting the
+ * so-called 'sysfs' (see also https://en.wikipedia.org/wiki/Sysfs) provided by Linux (a pseudo file system provided by
+ * the Linux kernel usually mounted at '/sys').
+ * <p/>
+ * A scan starts by inspecting the contents of the directory '/sys/class/tty'. This directory contains a symbolic link
+ * for every serial port style device that points to the device information provided by the sysfs in some subdirectory
+ * of '/sys/devices'.
+ * <p/>
+ * The scan considers only those serial ports for which the corresponding device file (in folder '/dev'; e.g.:
+ * '/dev/ttyUSB0') is both readable and writable, as otherwise the serial port cannot be used by any binding. For those
+ * serial ports, the scan checks whether the serial port actually originates from a USB device, by inspecting the
+ * information provided by the sysfs in the folder pointed to by the symbolic link.
+ * <p/>
+ * If the device providing the serial port is a USB device, information about the device (vendor ID, product ID, etc.)
+ * is collected from the sysfs and returned together with the name of the serial port in form of a
+ * {@link UsbSerialDeviceInformation}.
+ *
+ * @author Henning Sudbrock - initial contribution
+ */
+@Component(configurationPid = "discovery.usbserial.linux.sysfs.usbserialscanner")
```

Comment:
Good question. Why did I use this PID? Because the `UsbSerialDiscoveryService` has PID `discovery.usbserial` (which I chose to be consistent with mDNS and UPnP discovery services), and this service is intended to be plugged into the `UsbSerialDiscoveryService` - and so I put it into the 'discovery.usbserial' namespace.

Are there any guidelines for config PIDs used in Eclipse SmartHome? Or do you have a suggestion for the PID?

Reply:
Valid point. Giving the actual discovery service `UsbSerialDiscoveryService` the PID `discovery.XXX` is what all of our discovery services do.

However, I don't know a guideline on how the PID for services should look like which are a part of the actual discovery service.

Maybe you are right and its PID should be a child PID of `discovery.X`